### PR TITLE
Add Windows Builder

### PR DESCRIPTION
**Pull request by nof20**
Add experimental Windows Builder.

### CODE REVIEW

File: windows-builder/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+gcloud services enable compute.googleapis.com
+export PROJECT=$(gcloud info --format='value(config.project)')
+export MEMBER=$(gcloud projects describe $PROJECT --format 'value(projectNumber)')@cloudbuild.gserviceaccount.com
+gcloud projects add-iam-policy-binding $PROJECT --member=serviceAccount:$MEMBER --role='roles/compute.admin'
+gcloud projects add-iam-policy-binding $PROJECT --member=serviceAccount:$MEMBER --role='roles/iam.serviceAccountUser'
+```
+
+Clone this repository and build the builder:
+```sh
+gcloud builds submit --config=builder/cloudbuild.yaml builder/
```
Comment (ImJasonH) [inline windows-builder/README.md]: This is out of date now

File: windows-builder/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+Then, refer to the builder in your project's `cloudbuild.yaml`.  To spin up an ephemeral `n1-standard-1` VM on Compute Engine, simply provide the command you wish to execute.  Your Cloud Build workspace is synchronized to `C:\workspace` on the server first.
```
Comment (ImJasonH) [inline windows-builder/README.md]: Incomplete sentence "...on the server first"

File: windows-builder/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+```yaml
+steps:
+- name: 'gcr.io/$PROJECT_ID/windows-builder'
+  args: [ '<command goes here>' ]
```
Comment (ImJasonH) [inline windows-builder/README.md]: What kind of command goes here? Can you give an example that prints "hello world" or does the equivalent of `ls` or something?

File: windows-builder/README.md (hunk 1 above)
Comment (nof20) [reply windows-builder/README.md]: Clarified offline

File: windows-builder/README.md (hunk 2 above)
Comment (nof20) [reply windows-builder/README.md]: "Your Cloud Build workspace is synchronized to `C:\workspace` on the server first."

Meaning: we copy your workspace to C:\workspace on the server before the build runs.  Is that clearer?

File: windows-builder/README.md (hunk 3 above)
Comment (nof20) [reply windows-builder/README.md]: There are examples in the `images` directory.  I can call that out in the README to make it clearer.

File: windows-builder/README.md (hunk 2 above)
Comment (ImJasonH) [reply windows-builder/README.md]: Is "on the [thing] first" a phrase I've never heard before? A Britishism perhaps? I would have phrased that as something like:

Your Cloud Build workspace is synchronized to `C:\workspace` at server startup

I guess context clues make it clear what's being said either way, maybe consult with Nitya to be sure?

**Comment by filipesilva** [comment]
@nof20 I was trying out the windows builders and it looks like the command output log isn't displayed. 

`Executing command` shows up and that's all.

Following the instructions, and using the `hello.bat` example, I got a 15m build whose log looks like this:

```
... initialization logs ...
2018/11/08 16:05:30 Waiting for Windows password response
2018/11/08 16:05:51 Firewall rule already exists
2018/11/08 16:05:51 Set ingress firewall rule successfully
2018/11/08 16:05:51 Waiting for server to become available

Microsoft Windows [Version 10.0.17134.165]
The system cannot find the path specified.
2018/11/08 16:08:51 Copying workspace
#< CLIXML
<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>#< CLIXML
... message above repeated for several minutes ...
<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>2018/11/08 16:19:57 Executing command 
PUSH
DONE
```

**Comment by nof20** [comment]
Thanks for letting me know, I can reproduce your problem with the example image & will push a fix (docs got out of sync with code).

`scripts/start_windows.sh` simply starts a VM: it's up to you to take the IP address, username and password and put it into your app's `cloudbuild.yaml`; the main README gives brief information on the syntax expected.  The builder does not search for Windows VMs available in the project - I would expect people to need to segregate dev/prod workloads, for example.

It would be possible to have multiple different build configurations and direct builds to different servers.  You'd need a different `cloudbuild.yaml` for each one. 

This is pretty limited functionality right now - just experimental (as the README says).  I'd welcome suggestions or - even better - contributions.

**Comment by filipesilva** [comment]
Heya @nof20 thanks for the prompt reply! For context, I'm looking at CI alternatives for our Linux/Windows workloads in Angular and related repositories (CLI, Bazel rules, etc).

Now that you explained the persistent VM usage to me I get it and it's really obvious. I had read the `To use an existing Windows server instead` but for some reason it did not register in my head how the comment in `Performance` about the script related to it...

I can verify that with the `--command` arg I see the output! 

Using a new VM every time It's still a bit slow (the command seems to take about 15m total to run) and flaky (2 in 7 runs timed out in one of the VM operations). I haven't tried with the persistent VM but it should alleviate both of those problems.

The repeat lines still come up. I'm not sure what they mean. Do you know? 
```
Step #0: 2018/11/13 15:43:38 Waiting for server to become available
Step #0: 
Step #0: Microsoft Windows [Version 10.0.17134.165]
Step #0: The system cannot find the path specified.
Step #0: 2018/11/13 15:46:16 Copying workspace
Step #0: #< CLIXML
Step #0: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>#< CLIXML
Step #0: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>#< CLIXML
Step #0: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>#< CLIXML
# some 50 more I guess?

Step #0: Volume in drive C has no label.
Step #0: Volume Serial Number is F68B-42EF
Step #0: 
Step #0: Directory of c:\workspace
Step #0: 
Step #0: 11/13/2018 03:56 PM <DIR> .
Step #0: 11/13/2018 03:56 PM <DIR> ..
Step #0: 11/13/2018 03:46 PM 837 .bazelrc
Step #0: 11/13/2018 03:46 PM <DIR> .buildkite
Step #0: 11/13/2018 03:46 PM <DIR> .circleci
Step #0: 11/13/2018 03:47 PM 73 .clang-format
Step #0: 11/13/2018 03:47 PM 28 .gitignore
Step #0: 11/13/2018 03:47 PM 324 BUILD.bazel
Step #0: 11/13/2018 03:48 PM 804 cloudbuild.yaml
Step #0: 11/13/2018 03:48 PM <DIR> e2e
Step #0: 11/13/2018 03:49 PM 22,160 graph.png
Step #0: 11/13/2018 03:47 PM 1,096 LICENSE
Step #0: 11/13/2018 03:49 PM 1,621 package.json
Step #0: 11/13/2018 03:49 PM 1,147 postinstall.tsconfig.json
Step #0: 11/13/2018 03:48 PM 4,007 README.md
Step #0: 11/13/2018 03:49 PM 41 renovate.json
Step #0: 11/13/2018 03:56 PM <DIR> src
Step #0: 11/13/2018 03:48 PM 3,425 WORKSPACE
Step #0: 11/13/2018 03:56 PM 177,598 yarn.lock
Step #0: 13 File(s) 213,161 bytes
Step #0: 6 Dir(s) 20,062,982,144 bytes free
Finished Step #0
```

You have a table in the performance section that shows it took about 3m, but in my experiment just running the `dir` command, it took about 15m:
image

I'm guessing the repeat lines are actually related though, since during most of the 15m the lines keep coming in one by one. If I had to change a guess, I'd bet on the workspace copy being slow.

On the topic of suggestions, improving the workspace cleanup and concurrent builds would be chief among concerns for the CI use case. What do you think about using windows docker containers to run the commands themselves? 

Although at that point it does feel like a bit of a second level (start a VM, to start a docker container) into something that Cloud Build already does very well for linux containers.

**Pull request merged**